### ISSUE TITLE
[5.0.1] .with() doesn't work with Opaque types

### ISSUE THREAD

**Issue by adrienharnay** [gvergnaud/ts-pattern#167]
**Describe the bug**
Since 5.0.1, Opaque types don't work with `.with()`

**Code Sandbox with a minimal reproduction case**
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbziAhjAxgCzgXzgMyghDgHIYBnAWjDRgFMoA7UgbgChRJZE4B5WgEcArvVwEiJcgE8w9KvnoUYbduxiyxAZS4AbelphRgTAOZwAvGQCCpOAB8yAIVLt0EJsrgU9BoyfMrAFk0TAA6KBQmABNiAAoASjgAPjgABjCAVjgAfjgAImt8uAAuAqdilAo4HXB9Q2MzDnYAehb+AGl2VAxMOJ86v0bTBPY4ODCAd2AYPsL8gBo4RMtUgEZR8amZuYqllYtUgCZNifoAD0wUYWVgADd6ROaNOX4hUQaAyzeUEXoAHnmDnK+WSbg8Xgg7yGXyscRCswiUViIBWqQy2TyQLK+QqSSqPz+nyaajacA6fG6aCwcShvw+-jMp22sziQIJAnpMLM+yShzgGzGE2mrNxlWqnKJjNMvNWcBOQrCFyuNxg90eCVYQA

**Versions**
- TypeScript version: 5.1.3
- ts-pattern version: 5.0.1
- environment: ts playground

**Comment by gvergnaud** [comment]
Thanks for the report. Just out of curiosity, why would you want to pattern match on an opaque type? It seems to defeat the purpose of hiding the internal representation, no?

**Comment by adrienharnay** [comment]
We use Opaque types to add precision to some string types, e.g.

```
type EmailAddress = Opaque<string>;

Type PhoneNumber = Opaque<string>;

type DateIso = Opaque<string>;
```

Then in the codebase we create type checkers that return the Opaque type if the condition is met. All other code that will expect an `EmailAddress` won't accept a simple string. And we sometimes use `match()` to orchestrate such code.

**Comment by gvergnaud** [comment]
fixed in v5.0.4
